src/main/java/org/springframework/data/mapping/model/KotlinValueUtils.java

```diff
@@ -0,0 +1,397 @@
+		 * the value component and whether the component is a primitive.
+		 */
+		COPY {
+			@Override
+			public boolean shouldApplyBoxing(KType type, boolean optional, KParameter component) {
+
+				KType copyType = expandUnderlyingType(type);
+
+				if (copyType.getClassifier()instanceof KClass<?> kc && kc.isValue() || copyType.isMarkedNullable()) {
+					return true;
+				}
+
+				return false;
+			}
+
+			private static KType expandUnderlyingType(KType kotlinType) {
+
+				if (!(kotlinType.getClassifier()instanceof KClass<?> kc) || !kc.isValue()) {
+					return kotlinType;
+				}
+
+				List<KProperty<?>> properties = getProperties(kc);
+				if (properties.isEmpty()) {
+					return kotlinType;
+				}
+
+				KType underlyingType = properties.get(0).getReturnType();
+				KType componentType = ValueBoxing.resolveType(underlyingType);
+				KType expandedUnderlyingType = expandUnderlyingType(componentType);
+
+				if (!kotlinType.isMarkedNullable()) {
+					return expandedUnderlyingType;
+				}
+
+				if (expandedUnderlyingType.isMarkedNullable()) {
+					return kotlinType;
+				}
+
+				Type javaType = ReflectJvmMapping.getJavaType(expandedUnderlyingType);
+				boolean isPrimitive = javaType instanceof Class<?> c && c.isPrimitive();
+
+				if (isPrimitive) {
+					return kotlinType;
+				}
+
+				return expandedUnderlyingType;
+			}
+
+			static List<KProperty<?>> getProperties(KClass<?> kClass) {
+
+				if (kClass.isValue()) {
+
+					for (KCallable<?> member : kClass.getMembers()) {
+						if (member instanceof KProperty<?> kp) {
+							return Collections.singletonList(kp);
+						}
+					}
+				}
+
+				List<KProperty<?>> properties = new ArrayList<>();
+				for (KCallable<?> member : kClass.getMembers()) {
+					if (member instanceof KProperty<?> kp) {
+						properties.add(kp);
+					}
+				}
+
+				return properties;
+			}
+		};
+
+		public abstract boolean shouldApplyBoxing(KType type, boolean optional, KParameter component);
+
+	}
+
+	/**
+	 * Utility to represent Kotlin value class boxing.
+	 */
+	static class ValueBoxing {
+
+		private final KClass<?> kClass;
+
+		private final KFunction<?> wrapperConstructor;
+
+		private final boolean applyBoxing;
+
+		private final @Nullable ValueBoxing next;
+
+		/**
+		 * Creates a new {@link ValueBoxing} for a {@link KParameter}.
+		 *
+		 * @param rules boxing rules to apply.
+		 * @param parameter the copy or constructor parameter.
+		 */
+		@SuppressWarnings("ConstantConditions")
+		private ValueBoxing(BoxingRules rules, KParameter parameter) {
+			this(rules, parameter.getType(), (KClass<?>) parameter.getType().getClassifier(), parameter.isOptional());
+		}
+
+		private ValueBoxing(BoxingRules rules, KType type, KClass<?> kClass, boolean optional) {
+
+			KFunction<?> wrapperConstructor = null;
+			ValueBoxing next = null;
+			boolean applyBoxing;
+
+			if (kClass.isValue()) {
+
+				wrapperConstructor = kClass.getConstructors().iterator().next();
```

Comment:
@mp911de There's no guarantee that the first constructor is the primary one, nor that it has any parameters, so this can fail on something like this:
```
@JvmInline
value class Z(val value: String) {
    constructor() : this("Fail")
}
```
I recommend to use [`primaryConstructor`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect.full/primary-constructor.html) instead.